File: C#.Net/JDI/Web/Selenium/Elements/Complex/DropList.cs
using Epam.JDI.Core.Interfaces.Complex;
using System;
using Epam.JDI.Commons;
using Epam.JDI.Web.Selenium.Base;
using Epam.JDI.Web.Selenium.Elements.Base;
using OpenQA.Selenium;

namespace Epam.JDI.Web.Selenium.Elements.Complex
{
    public class DropList : DropList<IConvertible>, IDropList
    {
        public DropList() { }

        public DropList(By valueLocator) : base(valueLocator) { }

        public DropList(By valueLocator, By optionsNamesLocator, By allOptionsNamesLocator = null)
            : base(valueLocator, optionsNamesLocator, allOptionsNamesLocator) { }
    }

    public class DropList<TEnum> : MultiSelector<TEnum>, IDropList<TEnum>
        where TEnum : IConvertible
    {
        private readonly GetElementType _button = new GetElementType();

        public DropList() : this(null) { }

        public DropList(By valueLocator) : base(valueLocator)
        {
            InitActions();
        }
        
        public DropList(By valueLocator, By optionsNamesLocator, By allOptionsNamesLocator = null)
            : base(optionsNamesLocator, allOptionsNamesLocator)
        {
            InitActions();
            _button = new GetElementType(valueLocator);
        }

        private void InitActions()
        {
            SelectListNamesAction = (d, names) =>
            {
                if (names == null || names.Count == 0)
                    return;
                if (Button != null)
                {
                    ExpandNameAction(this, names[0]);
                    SelectListNamesAction(this, names);
                }
                else
                    names.ForEach(name => Selector.SelectByText(name));
            };
            SelectListIndexesAction = (d, indexes) =>
            {
                if (indexes == null || indexes.Count == 0)
                    return;
                if (Button != null)
                {
                    ExpandNumAction(this, indexes[0]);
                    SelectListIndexesAction(this, indexes);
                }
                else
                    indexes.ForEach(index => Selector.SelectByIndex(index));
            };
            ClearAction = d =>
            {
                if (Button != null)
                    ExpandNumAction(this, 1);
                ClearAction(this);
            };
            GetValueAction = d => GetTextAction(this);
        }

        protected Clickable Button => _button.Get(new Clickable(), WebAvatar);

        protected Action<DropList<TEnum>, string> ExpandNameAction = (d, name) =>
        {
            if (!d.DisplayedNameAction(d, name))
                d.Button.Click();
        };

        protected Action<DropList<TEnum>, int> ExpandNumAction = (d, index) =>
        {
            if (!d.DisplayedNumAction(d, index))
                d.Button.Click();
        };
        
        protected Func<DropList<TEnum>, string> GetTextAction = 
            d => d.WebElement.GetAttribute("value");
        
        
        public new void WaitDisplayed()
        {
            Button.WaitDisplayed();
        }
        
        public new void WaitVanished()
        {
            Button.WaitVanished();
        }

        public void Wait(Func<IWebElement, bool> resultFunc)
        {
            Button.Wait(resultFunc);
        }

        public T Wait<T>(Func<IWebElement, T> resultFunc, Func<T, bool> condition)
        {
            return Button.Wait(resultFunc, condition);
        }

        public void Wait(Func<IWebElement, bool> resultFunc, int timeoutSec)
        {
            Button.Wait(resultFunc, timeoutSec);
        }

        public T Wait<T>(Func<IWebElement, T> resultFunc, Func<T, bool> condition, int timeoutSec)
        {
            return Button.Wait(resultFunc, condition, timeoutSec);
        }

        public void SetAttribute(string attributeName, string value)
        {
            Button.SetAttribute(attributeName, value);
        }

        public string Text => Actions.GetText(d => GetTextAction(this));

        public string WaitText(string text)
        {
            return Actions.WaitText(text, d => GetTextAction(this));
        }

        public string WaitMatchText(string regEx)
        {
            return Actions.WaitMatchText(regEx, d => GetTextAction(this));
        }

        public new IWebElement WebElement => new WebElement(Locator).WebElement;

        public string GetAttribute(string name)
        {
            return Button.GetAttribute(name);
        }

        public void WaitAttribute(string name, string value)
        {
            Button.WaitAttribute(name, value);
        }
    }
}

File: jdi-uitest-web/src/main/java/com/epam/jdi/uitests/web/selenium/elements/complex/DropList.java
package com.epam.jdi.uitests.web.selenium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.complex.IDropList;
import com.epam.jdi.uitests.web.selenium.elements.GetElementType;
import com.epam.jdi.uitests.web.selenium.elements.base.BaseElement;
import com.epam.jdi.uitests.web.selenium.elements.base.Clickable;
import com.epam.jdi.uitests.web.selenium.elements.base.Element;
import com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.objects.JDropList;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.lang.reflect.Field;
import java.util.function.Function;

import static com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.WebAnnotationsUtil.findByToBy;
import static com.epam.jdi.uitests.web.selenium.elements.pageobjects.annotations.objects.FillFromAnnotationRules.fieldHasAnnotation;

/**
 * Select control implementation
 *
 * @author Alexeenko Yan
 * @author Belousov Andrey
 */
public class DropList<TEnum extends Enum> extends MultiSelector<TEnum> implements IDropList<TEnum> {
    private GetElementType button = new GetElementType();

    public DropList() {
        super();
    }

    public DropList(By valueLocator) {
        super(valueLocator);
    }

    public DropList(By valueLocator, By optionsNamesLocator) {
        this(valueLocator, optionsNamesLocator, null);
    }

    public DropList(By valueLocator, By optionsNamesLocator, By allOptionsNamesLocator) {
        super(optionsNamesLocator, allOptionsNamesLocator);
        this.button = new GetElementType(valueLocator, this);
    }

    public static void setUp(BaseElement el, Field field) {
        if (!fieldHasAnnotation(field, JDropList.class, IDropList.class)) {
            return;
        }
        ((DropList) el).setUp(field.getAnnotation(JDropList.class));
    }

    public DropList setUp(JDropList jDropList) {
        By root = findByToBy(jDropList.root());
        By list = findByToBy(jDropList.list());
        By value = findByToBy(jDropList.value());

        if (root != null) {
            Element el = new Element(root);
            el.setParent(getParent());
            setParent(el);
            setAvatar(root);
        }
        if (list != null) {
            this.allLabels = new GetElementType(list, this);
        }

        if(value != null) {
            this.button = new GetElementType(value, this);
        }
        return this;
    }

    protected Clickable button() {
        return button.get(Clickable.class);
    }

    protected void expandAction(String name) {
        if (!isDisplayedAction(name)) button().click();
    }

    protected void expandAction(int index) {
        if (!isDisplayedAction(index)) button().click();
    }

    @Override
    protected void selectListAction(String... names) {
        if (names == null || names.length == 0) {
            return;
        }
        if (button() != null) {
            expandAction(names[0]);
            super.selectListAction(names);
        } else
            for (String name : names)
                getSelector().selectByVisibleText(name);
    }

    @Override
    protected void selectListAction(int... indexes) {
        if (indexes == null || indexes.length == 0)
            return;
        if (button() != null) {
            expandAction(indexes[0]);
            super.selectListAction(indexes);
        } else
            for (int index : indexes)
                getSelector().selectByIndex(index);
    }

    @Override
    protected void clearAction() {
        if (button() != null)
            expandAction(1);
        super.clearAction();
    }

    @Override
    protected String getValueAction() {
        return getTextAction();
    }

    protected String getTextAction() {
        String getValue = getWebElement().getAttribute("value");
        String getText = getWebElement().getText();
        return getText.equals("") && getValue != null ? getValue : getText;
    }

    /**
     * Waits while Element becomes visible
     */
    @Override
    public void waitDisplayed() {
        button().waitDisplayed();
    }

    /**
     * Waits while Element becomes invisible
     */
    @Override
    public void waitVanished() {
        button().waitVanished();
    }

    public void wait(Function<WebElement, Boolean> resultFunc) {
        button().wait(resultFunc);
    }

    public <R> R wait(Function<WebElement, R> resultFunc, Function<R, Boolean> condition) {
        return (R) button().wait(resultFunc, condition);
    }

    public void wait(Function<WebElement, Boolean> resultFunc, int timeoutSec) {
        button().wait(resultFunc, timeoutSec);
    }

    public <R> R wait(Function<WebElement, R> resultFunc, Function<R, Boolean> condition, int timeoutSec) {
        return (R) button().wait(resultFunc, condition, timeoutSec);
    }

    /**
     * @param attributeName Specify attribute name
     * @param value         Specify attribute value
     *                      Sets attribute value for Element
     */
    public void setAttribute(String attributeName, String value) {
        button().setAttribute(attributeName, value);
    }

    /**
     * @return Get Element’s text
     */
    public final String getText() {
        return actions.getText(this::getTextAction);
    }

    /**
     * @param text Specify expected text
     * @return Wait while Element’s text contains expected text. Returns Element’s text
     */
    public final String waitText(String text) {
        return actions.waitText(text, this::getTextAction);
    }

    /**
     * @param regEx Specify expected regular expression Text
     * @return Wait while Element’s text matches regEx. Returns Element’s text
     */
    public final String waitMatchText(String regEx) {
        return actions.waitMatchText(regEx, this::getTextAction);
    }

    public WebElement getWebElement() {
        return new GetElementType(getLocator(), this).get(Clickable.class).getWebElement();
    }

    /**
     * Get element attribute
     *
     * @param name Specify name for attribute
     * @return Returns chosen attribute
     */
    public String getAttribute(String name) {
        return button().getAttribute(name);
    }

    /**
     * @param name  Specify attribute name
     * @param value Specify attribute value
     * Waits while attribute gets expected value. Return false if this not happens
     */
    public void waitAttribute(String name, String value) {
        button().waitAttribute(name, value);
    }
}